tests/unit/orchestrate.py

```diff
@@ -0,0 +1,450 @@
+    # run. We'll start by constructing a sequence that would run each test through to
+    # completion serially, one after the other.
+    test_sequence = []
+    counts = []
+    for index, test in enumerate(tests):
+        thread = _TestThread(test, name)
+        for count in itertools.count(1):  # pragma: NO BRANCH
+            # Pragma is required because loop never finishes naturally.
+            thread.go()
+            if thread.finished:
+                break
+
+        counts.append(count)
+        test_sequence += [index] * count
+
+    # Now we can take that initial sequence and generate all of its permutations,
+    # running each one to try to uncover concurrency bugs
+    sequences = iter(_permutations(test_sequence))
+
+    # We already tested the first sequence getting our counts, so we can discard it
+    next(sequences)
+
+    # Test each sequence
+    for test_sequence in sequences:
+        threads = [_TestThread(test, name) for test in tests]
+        try:
+            for index in test_sequence:
+                threads[index].go()
+
+            # Its possible for number of turns to vary from one test run to the other,
+            # especially if there is some undiscovered concurrency bug. Go ahead and
+            # finish running each test to completion, if not already complete.
+            for thread in threads:
+                while not thread.finished:
+                    thread.go()
+
+        except Exception:
+            # If an exception occurs, we still need to let any threads that are still
+            # going finish up. Additional exceptions are silently ignored.
+            for thread in threads:
+                thread.finish()
+            raise
+
+    return tuple(counts)
+
+
+_local = threading.local()
+
+
+class _Conductor:
+    """Coordinate communication between main thread and a test thread.
+
+    Two way communicaton is maintained between the main thread and a test thread using
+    two synchronized queues (`queue.Queue`) each with a size of one.
+    """
+
+    def __init__(self):
+        self._notify = queue.Queue(1)
+        self._go = queue.Queue(1)
+
+    def notify(self):
+        """Called from test thread to let us know it's finished or is ready for its next
+        turn."""
+        self._notify.put(None)
+
+    def standby(self):
+        """Called from test thread in order to block until told to go."""
+        self._go.get()
+
+    def wait(self):
+        """Called from main thread to wait for test thread to either get to the
+        next syncpoint or finish."""
+        self._notify.get()
+
+    def go(self):
+        """Called from main thread to tell test thread to go."""
+        self._go.put(None)
+
+
+_SYNCPOINTS = {}
+"""Dict[str, Dict[str, Set[int]]]: Dict mapping source fileneme to a dict mapping
+syncpoint name to set of line numbers where syncpoints with that name occur in the
+source file.
+"""
+
+
+def _get_syncpoints(filename):
```

Comment:
This is a very simplistic implementation that requires every pragma be located on a reachable line of code. This is probably sufficient. A more sophisticated algorithm that allowed pragmas to appear on their own lines is possible, but not trivial. See [this comment](https://github.com/googleapis/python-ndb/pull/709/files#r694361313).